### ISSUE TITLE
 Error: object 'data_survival_curve' not found 

### ISSUE THREAD

**Issue by alik33** [rstudio/shiny#1986]
This is probably problem of scoping (rather bug of scoping). I studied R Shiny scoping and I do not see that I do not follow the rules of scoping.

So I post it here. Let me know if you see it rather as bug of Shiny or there is some mistake made by me.

After running `runApp('script.R')`, I am getting the following error `Error: object 'data_survival_curve' not found`. I run debug in RStudio and on line nr 60 this variable is created and it exists till the moment when error comes.

script.R file:
```
library(shiny)
library(survival)
library(survminer)
library(directlabels)

data <- read.csv('dataset.csv', header = TRUE, sep = ",", fileEncoding="UTF-8")
unique_transplant_years_decreasing <- as.numeric(sort(unique(c(data$transplant_year)), decreasing = TRUE))

krivkaPreziti <- sidebarLayout(
  # all inputs for graph krivka preziti
  sidebarPanel(
    sliderInput("krivka_preziti_input_years", 'Years:', 
                min = unique_transplant_years_decreasing[length(unique_transplant_years_decreasing)],
                max = unique_transplant_years_decreasing[1],
                value = c(unique_transplant_years_decreasing[length(unique_transplant_years_decreasing)],
                          unique_transplant_years_decreasing[1]),
                step = 1),
    numericInput('krivka_preziti_input_seskupit_po', 'Group by (years):',
                 value = 0,
                 min = 0),
    checkboxInput('krivka_preziti_input_facet', 'Facet', value = FALSE),
    width = 3
  ),
  
  # Create a spot for bar plot
  mainPanel(
    h2('Survival curve'),
    br(),
    plotOutput('krivka_preziti', height = "750px"),
    width = 12
  )
)



panelAnalyzaPreziti <- tabPanel(
  'Survival analysis',
  krivkaPreziti 
)


ui <- navbarPage(
  title = "Application",
  panelAnalyzaPreziti
)


server <- shinyServer(
  function(input, output, session)
  {
    ################################# TAB 4 ########################################
    output$krivka_preziti <- renderPlot(
      {
        krivka_year_bottom <- input$krivka_preziti_input_years[1]
        krivka_year_top <- input$krivka_preziti_input_years[2]
        krivka_seskupit_po <- input$krivka_preziti_input_seskupit_po
        
        # data which fit the range of selected years
        # data which meet the condition that survival_time is not NA
        data_survival_curve <- data[data$transplant_year %in% seq(krivka_year_bottom, krivka_year_top) &
                                      !is.na(data$survival_time) &
                                      data$survival_time >= 0,]
        
        
        # if seskupit_po != 0, then cut 
        if(krivka_seskupit_po != 0) {
          data_survival_curve$time_period <- cut(as.numeric(data_survival_curve$transplant_year),
                                                   seq(krivka_year_bottom, krivka_year_top, krivka_seskupit_po),
                                                   include.lowest = T)
          data_survival_curve <- data_survival_curve[!is.na(data_survival_curve$time_period),]
          data_survival_curve$time_period <- as.factor(data_survival_curve$time_period)
        }
        else {
          data_survival_curve$time_period = data_survival_curve$transplant_year
        }
        
        # validate number of rows of data set > 0
        shiny::validate(
          need(nrow(data_survival_curve) > 0, 'Broader your input')
        )
        
        surv_obj <- Surv(data_survival_curve$survival_time/365,data_survival_curve$patient_died)
        fit <- survfit(surv_obj ~ time_period, data = data_survival_curve)
        
        krivka_preziti_plt <- ggsurvplot(fit,
                                         linetype = c('solid'),
                                         ggtheme = theme_bw(),
                                         surv.scale = 'percent',
                                         xlab = 'Roky',
                                         ylab = '%',
                                         censor = FALSE,
                                         break.x.by = 1,
                                         break.y.by = 0.1)
        plot2 <- krivka_preziti_plt + geom_dl(aes(label = time_period), method = list("last.points"), cex = 0.8)
        
        plot2
        
      }
    )
  }
)

shinyApp(
  ui = ui,
  server = server
)
```

Here is the data set that I am using:
[dataset.csv](https://uploadfiles.io/ry2gj)

**Comment by wch** [comment]
Can you provide a minimal example? A good way to proceed is to remove things bit by bit until you end up with a very small example that exhibits the problem.

**Comment by alik33** [comment]
Sure. This does not exactly capture my problem from the original topic message but it also gives the same error and I do not understand why. We can slowly move forward to the original problem that I faced.

`app.R`:
```
library(shiny)
library(survival)
library(survminer)
library(directlabels)

data_survival_curve <- read.csv('dataset.csv', header = TRUE, sep = ",", fileEncoding="UTF-8")

ui <- fluidPage(
  mainPanel(
    h2('Survival curve'),
    br(),
    plotOutput('survival_curve', height = "750px"),
    width = 12
  )
)


server <- shinyServer(
  function(input, output, session)
  {
    ################################# TAB 4 ########################################
    output$survival_curve <- renderPlot(
      {
        data_survival_curve$time_period <- data_survival_curve$transplant_year
        
        surv_obj <- Surv(data_survival_curve$survival_time/365,data_survival_curve$patient_died)
        fit <- survfit(surv_obj ~ time_period, data = data_survival_curve)
        
        krivka_preziti_plt <- ggsurvplot(fit,
                                         linetype = c('solid'),
                                         ggtheme = theme_bw(),
                                         surv.scale = 'percent',
                                         xlab = 'Years',
                                         ylab = '%',
                                         censor = FALSE,
                                         break.x.by = 1,
                                         break.y.by = 0.1)
        plot2 <- krivka_preziti_plt + geom_dl(aes(label = time_period), method = list("last.points"), cex = 0.8)
        
        plot2
        
      }
    )
  }
)

shinyApp(
  ui = ui,
  server = server
)
```
Now if I navigate to the directory where this `app.R` in R and I run: `runApp()`, it shows the error that is written in the title of this issue.

**Comment by jcheng5** [comment]
I suspect it's a problem with ggsurvplot and nonstandard evaluation, can you verify that it _doesn't_ happen if you use ggplot2 directly?

**Comment by alik33** [comment]
@jcheng5 sorry, but what do you exactly mean by verifying it by using ggplot2 directly?

**Comment by jcheng5** [comment]
I investigated--it's not technically a Shiny bug but I found a workaround (from https://github.com/kassambara/survminer/issues/9). The issue basically is that `survfit` uses nonstandard evaluation to capture _not_ the model formula and data, but rather, the _expression_ that the caller used to specify the model formula and data. These expressions can only be reliably evaluated using the environment they were originally defined in--the `parent.frame()` of the `survfit` call. However, `survfit` doesn't save a reference to this environment AFAICT. Anyway, `ggsurvplot` needs the values, so it just evaluates those expressions in (essentially) the global environment.

The workaround is to replace this line:

```r
fit <- survfit(surv_obj ~ time_period, data = data_survival_curve)
```

with this:

```r
fit <- do.call(survfit, list(surv_obj ~ time_period, data = data_survival_curve))
```

A quirk of `do.call` is that the arguments of the call are essentially made environment-independent.
